common-tools/clas-tracking/src/main/java/org/jlab/clas/tracking/kalmanfilter/AMeasVecs.java

```diff
@@ -33,9 +35,57 @@ public void setMeasVecs(List<Surface> measSurfaces) {
             mvec.skip = mvec.surface.passive;
             mvec.hemisphere = measSurfaces.get(i).hemisphere;
             measurements.add(mvec);
+            if(measSurfaces.get(i).type == Type.LINEDOCA) {
+            	mvec.region = measSurfaces.get(i).region;
+            	mvec.sector = measSurfaces.get(i).getSector();
+            	mvec.superlayer = measSurfaces.get(i).getSuperLayer();
+            	mvec.layer = measSurfaces.get(i).getLayer();
+            }
         }
     }
     
+    // For DC, there could be two hits at a measurement
+    public double[] dhDoca(int k, StateVec stateVec) {  
+    	double value[] = {Double.NaN, Double.NaN};
+    	
+    	Surface surf = this.measurements.get(stateVec.k).surface;
+    	Point3D point = new Point3D(stateVec.x, stateVec.y, stateVec.z);
+    	double h = hDoca(point, surf.wireLine[0]);
+
+    	double signMeas = 1;
+    	double sign = 1;
+    	if(surf.doca[1]!=-99 || !(Math.abs(surf.doca[0])<0.5 && surf.doca[1]==-99 ) ) { //use LR only for double hits or large enough doca for one hit
+    		signMeas = Math.signum(surf.doca[0]);
+    		sign = Math.signum(h);
+    	} else {
+    		signMeas = Math.signum(h);
+    		sign = Math.signum(h);
+    	}
+
+    	value[0] = signMeas*Math.abs(surf.doca[0]) - sign*Math.abs(h);
+    	
+        //USE THE DOUBLE HIT
+        if(surf.doca[1]!=-99) { 
+            h = hDoca(point, surf.wireLine[1]);
+
+            signMeas = Math.signum(surf.doca[1]);
+            sign = Math.signum(h);
+            
+            value[1]= signMeas*Math.abs(surf.doca[1]) - sign*Math.abs(h);
+        } 
+    	
+    	return value;
+    }
+    
+    // Return a signed doca for DC
+    public double hDoca(Point3D point, Line3D wireLine) {              
```

Comment:
Parameters of the method dh is k for measurement-site number and stateVec for state vector at the current site. However, for forward tracking, since existence of double hits at a measurement site, one more parameter should be added to tell "h" is calculated for which hit. 
Additionally, for cases of double hits, KF filter is applied one by one for HB tracking. So there is an intermediate state after KF filter is applied for the first hit. But we do not save the intermediate state. For such case, Point3D is more comfortably used as parameter for h calculation of the second hit.
Therefore, it cannot simply unify h calculation for central and forward tracking.
Sure, we can force it to be unified, but it is far-fetched in my opinion.

Reply:
I agree it is not a trivial restructuring but it is anyway needed..